cpp/rng/rng.cpp

```diff
@@ -0,0 +1,224 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#include "rng.h"
+#include "../util/error.h"
+
+#include <chrono>
+#include <random>
+#include <sstream>
+#include <string.h>
+
+namespace openage {
+namespace rng {
+
+rng::rng(uint64_t v1) {
+	this->seed(v1);
+}
+
+rng::rng(const void *data, size_t len) {
+	this->seed(data, len);
+}
+
+rng::rng(const std::string &instr) {
+	this->from_string(instr);
+}
+
+rng::rng(std::istream &instream) {
+	this->from_stream(instream);
+}
+
+void rng::seed(uint64_t val) {
+	this->seed(&val, sizeof(val));
+}
+
+/*
+ * murmurhash3 bit avalancher - output is uniformly distributed
+ * and changing one bit of input results in ~1/2 of output bits changed
+ * This function is a good way of mixing up an input,
+ * or mixing together two results with hash64(val1 ^ val2)
+ */
+static uint64_t hash64(uint64_t val) {
+	val ^= val >> 33;
+	val *= 0xff51afd7ed558ccd;
+	val ^= val >> 33;
+	val *= 0xc4ceb9fe1a85ec53;
```

Comment:
unfortunately the wikipedia article doesn't mention those magic numbers, it uses other ones...